index.js

```diff
@@ -224,6 +224,43 @@ export function execaSync(file, args, options) {
 	};
 }
 
+export function $(templatesOrOptions, ...expressions) {
+	if (Array.isArray(templatesOrOptions)) {
+		const [file, ...args] = parseCommand(join(templatesOrOptions, expressions));
+		return execa(file, args);
+	}
+
+	return (templates, ...expressions) => {
+		const [file, ...args] = parseCommand(join(templates, expressions));
+		return execa(file, args, templatesOrOptions);
+	};
+}
+
+$.sync = (templatesOrOptions, ...expressions) => {
+	if (Array.isArray(templatesOrOptions)) {
+		const [file, ...args] = parseCommand(join(templatesOrOptions, expressions));
+		return execaSync(file, args);
+	}
+
+	return (templates, ...expressions) => {
+		const [file, ...args] = parseCommand(join(templates, expressions));
+		return execaSync(file, args, templatesOrOptions);
+	};
+};
+
+function join(templates, expressions) {
+	let command = '';
+	for (const [i, template] of templates.entries()) {
+		command += `${template}${parseExpression(expressions[i])}`;
+	}
+
+	return command;
+}
+
+function parseExpression(expression) {
+	return Array.isArray(expression) ? expression.join(' ') : expression ?? '';
```

Comment:
The current logic would remove whitespace escaping. For example:

```js
$`file ${'Hello world'}` -> execa('file', ['Hello', 'world'])
```

instead of:

```js
$`file ${'Hello world'}` -> execa('file', ['Hello world'])
```

Also `parseCommand()` allow backslashes as escape characters, even though they are not needed there, so this function might not fit there. Calling just `string.split()` should be enough actually.

We also need to take into account that `template` might be `undefined`, when the template string contains invalid backlash sequences like an isolated `\u`. 

One possible implementation could be:

```js
const $ = (templates, ...expressions) => {
	const {file, args} = parseTemplates(templates, expressions);
	return execa(file, args);
};

const parseTemplates = (templates, expressions) => {
	const [file, ...args] = templates.flatMap((template, index) => parseTemplate(template, index, templates, expressions));
	return {file, args};
};

const parseTemplate = (template, index, templates, expressions) => {
	const templateString = template ?? templates.raw[index];
	const templateTokens = templateString.split(SPACES_REGEXP).filter(Boolean);

	if (index === expressions.length) {
		return templateTokens;
	}

	const expression = expressions[index];
	return Array.isArray(expression) ? [...templateTokens, ...expression.map(String)] : [...templateTokens, String(expression)];
};

const SPACES_REGEXP = / +/g;
```

I left the sync logic and the options binding logic out for the sake of the example.

This produces the following result:

```js
$`file one two` -> execa(file, ['one', 'two'])
$`file  one  two` -> execa(file, ['one', 'two'])
$`file ${'one'} ${'two'}` -> execa(file, ['one', 'two'])
$`file ${0}` -> execa(file, ['0'])
$`file "one two"` -> execa(file, ['"one', 'two"'])

$`file ${'one two'}` -> execa(file, ['one two'])
$`file ${['one', 'two three']}` -> execa(file, ['one', 'two three'])
$`file \n` -> execa(file, ['\n'])
$`file ${'\n'}` -> execa(file, ['\n'])
$`file \u` -> execa(file, ['\\u'])
$`file ${undefined}` -> execa(file, ['undefined'])
$`file one\\ two` -> execa(file, ['one\\', 'two'])
$`file ${'"one two"'}` -> execa(file, ['"one two"'])
```

Compared to the current implementation:

```js
$`file one two` -> execa(file, ['one', 'two'])
$`file  one  two` -> execa(file, ['one', 'two'])
$`file ${'one'} ${'two'}` -> execa(file, ['one', 'two'])
$`file ${0}` -> execa(file, ['0'])
$`file "one two"` -> execa(file, ['"one', 'two"'])

$`file ${'one two'}` -> execa(file, ['one', 'two'])
$`file ${['one', 'two three']}` -> execa(file, ['one', 'two', 'three'])
$`file \n` -> execa(file, [])
$`file ${'\n'}` -> execa(file, [])
$`file \u` -> execa(undefined, [])
$`file ${undefined}` -> execa(file, [])
$`file one\\ two` -> execa(file, ['one two'])
$`file ${'"one two"'}` -> execa(file, ['"one', 'two"'])
```

What do you think?

Reply:
Ah! Okay these examples are great and really drive home your point in our original discussion:

> Which means that quotes and backslashes do not need to be interpreted as escaping characters, since ${...} already fulfills escaping needs. They would be interpreted as regular characters instead.

I'm totally on board with this approach! Also, thanks for the sample implementation. It works great!
https://github.com/sindresorhus/execa/pull/510/commits/8aa4e0714cb98198641eed48b9aeb7266f6c4346